### PR TITLE
Updating AWS Session creation to be able to use AWS_CA_BUNDLE environment variable

### PR DESCRIPTION
**Pull request by bhks**
Updating AWS Session creation to be able to use AWS_CA_BUNDLE environment variable 

The Issue was that the current code was not able to utilize custom AWS_CA_BUNDLE to create aws session.

Testing :

ON an existing EKS cluster I updated it to have  encryption provider enabled and associated. Then Applied the kubectl commands to test that the encryption and decryption is working.

```
bash-3.2$ MASTER_KEY_ARN=$(aws kms create-key --query KeyMetadata.Arn --output text)
bash-3.2$ aws kms create-alias \
>       --alias-name alias/k8s-master-key \
>       --target-key-id $(echo $MASTER_KEY_ARN | cut -d "/" -f 2)
bash-3.2$

aws eks associate-encryption-config \
    --cluster-name sbhagwat-test-inplace \
    --encryption-config '[{"resources":["secrets"],"provider":{"keyArn":"XXXXXX"}}]' \
 --region us-west-2


kubectl create secret \
        generic test-creds \
        --from-file=test-creds=/tmp/test-creds \
        --namespace secretslab


sh-4.2$ kubectl create ns secretslab
namespace/secretslab created
sh-4.2$ 
sh-4.2$ echo -n "am i safe?" > ./test-creds
sh: ./test-creds: Permission denied
sh-4.2$ echo -n "am i safe?" > /tmp/test-creds
sh-4.2$ 
sh-4.2$ kubectl create secret \
>         generic test-creds \
>         --from-file=test-creds=/tmp/test-creds \
>         --namespace secretslab
secret/test-creds created
sh-4.2$ 
sh-4.2$ 
sh-4.2$ kubectl get secret test-creds \
>   -o jsonpath="{.data.test-creds}" \
>   --namespace secretslab | \
>   base64 --decode
am i safe?sh-4.2$ 
sh-4.2$   

```

### CODE REVIEW

File: pkg/cloud/cloud.go [hunk 1]
```diff
@@ -33,9 +33,15 @@ func New(region, kmsEndpoint string, qps, burst int) (*AWSKMS, error) {
 		CredentialsChainVerboseErrors: aws.Bool(true),
 		Endpoint:                      aws.String(kmsEndpoint),
 	}
+
+	sess, err := session.NewSession()
```
Comment (jyotimahapatra) [inline pkg/cloud/cloud.go:37]: is this similar to pattern followed in https://github.com/aws/aws-sdk-go/blob/main/aws/session/custom_ca_bundle_test.go#L52-L59 ?

Comment (bhks) [reply pkg/cloud/cloud.go:37]: That was the previous setup and it failed

Comment (jyotimahapatra) [reply pkg/cloud/cloud.go:37]: Does that mean `sess.HttpClient` overides `cfg.HttpClient` ? From https://github.com/aws/aws-sdk-go/blob/main/aws/session/session.go#L593 i see the CA bundle being used for the `cfg.HttpClient`

Comment (jyotimahapatra) [reply pkg/cloud/cloud.go:37]: is it possible to add a unit test similar to https://github.com/aws/aws-sdk-go/blob/main/aws/session/custom_ca_bundle_test.go#L52-L59 to make sure a future change doesn't revert this back?

Comment (jyotimahapatra) [reply pkg/cloud/cloud.go:37]: is it possible to add a unit test similar to https://github.com/aws/aws-sdk-go/blob/main/aws/session/custom_ca_bundle_test.go#L52-L59 to make sure a future change doesn't revert this back?

Comment (jyotimahapatra) [reply pkg/cloud/cloud.go:37]: ok nvm my previous qs.. if cfg is not provided then the session.HttpClient is used https://github.com/aws/aws-sdk-go/blob/9e261711fdee3b565ef0cb4f94a28610decaa0e0/aws/config.go#L505-L507 
Wondering how changing the HttpClient frm cfg to session helps.

File: pkg/cloud/cloud.go [hunk 2]
```diff
@@ -15,12 +15,12 @@ package cloud
 
 import (
 	"fmt"
+	"sigs.k8s.io/aws-encryption-provider/pkg/httputil"
```
Comment (micahhausler) [inline pkg/cloud/cloud.go:18]: Can you either put this back in the 3rd party import group or put it on in its own group?

File: pkg/cloud/cloud.go (hunk 1 above)
Comment (micahhausler) [reply pkg/cloud/cloud.go:37]: Looking at the docs for [Custom CA Bundle](https://docs.aws.amazon.com/sdk-for-go/api/aws/session/#:~:text=Custom%20CA%20Bundle) it states:
> The Session option CustomCABundle is also available when creating sessions to also enable this feature. CustomCABundle session option field has priority over the AWS_CA_BUNDLE environment variable, and will be used if both are set.
> Setting a custom HTTPClient in the aws.Config options will override this setting. To use this option and custom HTTP client, the HTTP client needs to be provided when creating the session. Not the service client.

This code LGTM

**Review by micahhausler — APPROVED** [review]
Code change LGTM. Just one nit on the imports

File: pkg/cloud/cloud.go (hunk 2 above)
Comment (bhks) [reply pkg/cloud/cloud.go:18]: Yes , I saw that later when I made the pull request I will correct it soon .

Comment (bhks) [reply pkg/cloud/cloud.go]: updated the code with right place of imports

File: pkg/cloud/cloud.go (hunk 1 above)
Comment (jyotimahapatra) [reply pkg/cloud/cloud.go:37]: Makes sense to me now. I would prefer if we could attempt a unit test instead of testing it manually on a cluster

Comment (bhks) [reply pkg/cloud/cloud.go:37]: updated with unit tests @jyotimahapatra

File: pkg/cloud/cloud_test.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	}
+}
+
+
+func TestNewSessionClientWithEnv(t *testing.T) {
+	tempFile, err := createTmpFile(TLSBundleCert)
+	os.Setenv("AWS_CA_BUNDLE", tempFile)
+	kmsObjet, err := New("us-west-2", "https://kms.us-west-2.amazonaws.com", 15, 5);
+	if kmsObjet == nil  || err !=nil {
+		t.Fatalf("Failed to create object with error (%v)", err)
```
Comment (jyotimahapatra) [inline pkg/cloud/cloud_test.go:39]: nit: use assert instead

File: pkg/cloud/cloud_test.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+
+func createTmpFile(b []byte) (string, error) {
+	bundleFile, err := ioutil.TempFile(os.TempDir(), "aws-sdk-go-session-test")
+	if err != nil {
+		return "", err
+	}
+
+	_, err = bundleFile.Write(b)
+	if err != nil {
```
Comment (jyotimahapatra) [inline pkg/cloud/cloud_test.go:51]: nit: use `assert`

**Review by jyotimahapatra — APPROVED** [review]

File: pkg/cloud/cloud_test.go (hunk 3 above)
Comment (bhks) [reply pkg/cloud/cloud_test.go:39]: If you look at all other files we have not been using assert

Comment (bhks) [reply pkg/cloud/cloud_test.go]: Anyway updated as per requested

File: pkg/cloud/cloud_test.go (hunk 4 above)
Comment (bhks) [reply pkg/cloud/cloud_test.go:50]: Done

**Review by jyotimahapatra — APPROVED** [review]
Thank you !

File: pkg/cloud/cloud_test.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+func TestNewSessionClientWithoutEnv(t *testing.T) {
+	kmsObjet, err := New("us-west-2", "https://kms.us-west-2.amazonaws.com", 15, 5);
+	assert.NoError(t, err, "Failed to create object with error (%v)", err)
+	assert.NotNil(t, kmsObjet, "Failed to create object with error (%v)", err)
+}
+
+
+func TestNewSessionClientWithEnv(t *testing.T) {
+	tempFile, err := createTmpFile(TLSBundleCert)
+	os.Setenv("AWS_CA_BUNDLE", tempFile)
```
Comment (micahhausler) [inline pkg/cloud/cloud_test.go:37]: 2 nits:

* Can you handle a possible err here?
* Can you unset this after your test so other tests don't have the variable set? 
```
t.Cleanup(func(){
  os.Unsetenv("AWS_CA_BUNDLE")
})
```

**Review by micahhausler — CHANGES_REQUESTED** [review]

File: pkg/cloud/cloud_test.go (hunk 5 above)
Comment (bhks) [reply pkg/cloud/cloud_test.go:37]: oops Thanks sure

Comment (bhks) [reply pkg/cloud/cloud_test.go:38]: updates the code

**Review by micahhausler — APPROVED** [review]
/lgtm
/approve

**Comment by micahhausler** [comment]
No change required in this PR, but we should probably rely on Go modules

/lgtm
/approve

**Pull request merged**